pkg/subvolume/subvolume.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type fsStruct struct {
+	Name string
+	data []string
+}
+
+type subvolume struct {
+	fs     string
+	svg    string
+	subvol string
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, clusterNamespace, operatorNamespace, flag string) {
+
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(context.Background(), v1.ListOptions{})
+	if err != nil {
+		fmt.Printf("Error fetching PVs: %v\n", err)
+		os.Exit(1)
+	}
+	var subvolumeName []string
+	for _, pv := range pvList.Items {
+		subvolumeName = append(subvolumeName, pv.Spec.CSI.VolumeAttributes["subvolumeName"])
+	}
+
+	sv := getCSIRef(ctx, clientsets, clusterNamespace, operatorNamespace, flag)
+
+	// if stale flag is passed, list only the stale subvolumes
+	if flag == "stale" {
+		// Check whether difference is of the two lists is measurable
+		diff := getDifference(sv, subvolumeName)
+		if diff == nil {
+			logging.Info("No stale volumes found")
+		} else {
+			logging.Info("%s", diff)
+		}
+	} else {
+		logging.Info("%s", sv)
+	}
+}
+
+func getDifference(subvol []subvolume, k8ref []string) []subvolume {
+
+	mb := make(map[string]struct{}, len(k8ref))
+	for _, x := range k8ref {
+		mb[x] = struct{}{}
+	}
+	var diff []subvolume
+	for _, x := range subvol {
+		if _, found := mb[x.subvol]; !found {
+			// reference present in ceph but not in k8s
+			st := subvolume{x.fs, x.subvol, x.svg}
+			diff = append(diff, st)
+		}
+	}
+	if diff == nil {
+		logging.Fatal(fmt.Errorf("No stale Volume found"))
+	}
+
+	return diff
+}
+
+func getCSIRef(ctx context.Context, clientsets *k8sutil.Clientsets, clusterNamespace, operatorNamespace, flag string) []subvolume {
+
+	fsList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "ls", "--format", "json"}, clusterNamespace, operatorNamespace, true, false)
+	if fsList == "" {
+		logging.Fatal(fmt.Errorf("failed to get filesystem"))
+	}
+
+	fsstruct := unMarshaljson(fsList)
+	var svList string
+	var staleVolume []subvolume
+
+	for _, fs := range fsstruct {
+		svgList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolumegroup", "ls", fs.Name, "--format", "json"}, clusterNamespace, operatorNamespace, true, false)
+		if strings.TrimSpace(svgList) == "[]" {
+			logging.Fatal(fmt.Errorf("failed to get subvolumegroup"))
+		}
+		subvolg := unMarshaljson(svgList)
+		for _, svg := range subvolg {
+			svList = exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "ls", fs.Name, svg.Name}, clusterNamespace, operatorNamespace, true, false)
+			if strings.TrimSpace(svList) == "[]" {
+				logging.Info("No subvolume found for fs %s svg %s", fs, svg)
+			}
+			subvol := unMarshaljson(svList)
+
+			// append the subvolume which doesn't have any snapshot attached to it.
+			for _, sv := range subvol {
```

Comment:
If there are more than a few subvolumes, this could take a while to complete since each subvolume requires a new ceph command to be executed. Could we immediately print the subvolumes as they are being evaluated? Then the user can see output as it is progressing. For easier processing, we may also want to provide an option to output json format and include `stale: true|false` property so they can automate the process better. But perhaps the json option could be a follow-up PR.

Reply:
we can't print it in case of stale volumes at least, as we have to compare it with k8s ref and then print the diff.

Reply:
As in the separate comment, let's find a way to print it immediately.